### PR TITLE
Add new difficulty level for playing vs. computer, timed, chat

### PR DESCRIPTION
**Pull request by ragaeeb**
### Engine Choice: Stockfish 18

After researching the available chess engines for browser/TypeScript use, **Stockfish** is the clear winner:

| Engine | Notes |
|--------|-------|
| **Stockfish (WASM)** | Gold standard. Used by Lichess. Strongest open-source engine. Runs in-browser via WebAssembly. |
| `@lichess-org/stockfish-web` | Requires SharedArrayBuffer + CORS headers. More complex setup. |
| `chess.js` | Move validation only, not an engine. Already used in the project. |
| Custom minimax | Much weaker, harder to calibrate difficulty. |

I chose the **`stockfish` npm package (v18.0.5)** with the **lite single-threaded WASM build** (~7MB) — it's strong enough to crush any human player while being lightweight for browser delivery.

### Difficulty Levels

Matching the existing puzzle difficulty system (`beginner` / `intermediate` / `advanced`), each level maps to Stockfish's UCI parameters:

| Difficulty | ELO | Skill Level | Depth | Move Time |
|------------|-----|-------------|-------|-----------|
| **Beginner** | ~800 | 0 | 1 | 200ms |
| **Intermediate** | ~1500 | 10 | 8 | 500ms |
| **Advanced** | ~2500 | 20 | 15 | 1000ms |

### Files Created/Modified

1. **`src/types/computer.ts`** — `ComputerDifficulty` type and `DIFFICULTY_CONFIGS` with engine parameters per level
2. **`src/lib/engine.ts`** — `StockfishEngine` class wrapping the Stockfish Web Worker with UCI protocol communication
3. **`src/hooks/useStockfish.ts`** — React hook that manages the engine lifecycle and exposes `getBestMove(fen)`
4. **`src/app/game/computer/page.tsx`** — Replaced the random-move AI with Stockfish. Added an animated difficulty selection screen with card UI showing ELO ratings, descriptions, and icons
5. **`package.json`** — Added `stockfish` dependency and `postinstall` script to copy WASM files to `public/`
6. **`.gitignore`** — Added `/public/stockfish/` (generated at install time from node_modules)

### UX Flow

1. Player navigates to "Play vs. Computer"
2. Animated difficulty selection screen appears with three options
3. Player picks a difficulty, engine loads via WebAssembly in a Web Worker
4. Top status badge shows engine state (loading, thinking, ready)
5. "New Game" button lets the player return to difficulty selection at any time

**Chat feature implementation — completed TDD style with 25 new tests:**

### New files created:
| File | Purpose |
|---|---|
| `src/types/chat.ts` | `ChatMessage` type definition |
| `src/lib/chatApi.ts` | Client-side `sendChatMessage` function |
| `src/lib/chatApi.test.ts` | 3 tests for the API client |
| `src/hooks/useChat.ts` | Chat state management hook |
| `src/hooks/useChat.test.ts` | 9 tests for the hook |
| `src/components/ChatPanel.tsx` | Collapsible chat UI panel |
| `netlify/functions/chat.ts` | Server-side chat handler |
| `netlify/functions/tests/chat.test.ts` | 13 tests for the server handler |
| `src/app/.netlify/functions/chat/route.ts` | Next.js dev route proxy |

### Files modified:
| File | Change |
|---|---|
| `src/types/socket.ts` | Added `CHAT` event |
| `src/hooks/useGameSubscription.ts` | Added `onChatMessage` binding via `callbacksRef` |
| `src/hooks/useGameNetwork.ts` | Added `onChatMessage` to `GameNetworkCallbacks` |
| `src/app/game/[gameId]/page.tsx` | Wired chat hook, panel, and unread tracking |
| `package.json` | Updated test scripts |

### Key adaptations from the proposal:
1. **Used `getServerPusher()`** instead of inline Pusher — follows existing singleton pattern
2. **Used existing `checkRateLimit()`** (30/min) instead of custom Redis per-second logic
3. **Added `verifyPlayerSignature()`** — the proposal omitted session auth
4. **Used `getGameRecord()`** from `gameStore.ts` instead of raw Redis
5. **Removed `useCallback`** — React 19 compiler handles memoization
6. **Chat binding uses `callbacksRef`** pattern — consistent with our subscription stability fix
7. **Restructured `useChat` API** with `createSendMessage(playerId, playerSig)` factory — cleanly resolves the circular dependency where `receiveMessage` is needed before identity is available

**Server (`netlify/functions/tests/chat.test.ts`) — 8 new tests:**
- 405 for non-POST/non-OPTIONS methods (GET, PUT, etc.)
- 400 for malformed JSON body
- 400 when `id` field is missing
- 400 when `playerSig` field is missing
- Boundary: text at exactly 200 characters succeeds
- Both players can chat in the same game (cross-chat)
- Client-provided message `id` is preserved in the broadcast
- Broadcast targets the correct `private-game-{id}` channel

**Hook (`src/hooks/useChat.test.ts`) — 6 new tests:**
- `sending` resets to `false` after successful send
- `error` clears on the next successful send
- Non-`Error` thrown objects produce a fallback error message
- `playerSig` null (but `playerId` present) blocks sending
- Messages under `MAX_LOCAL_MESSAGES` are not trimmed
- `createSendMessage` returns a callable function

**API (`src/lib/chatApi.test.ts`) — 4 new tests:**
- Successful response resolves to `undefined` (no throw)
- `Content-Type` header is set to `application/json`
- Error JSON with no `error` field falls back to default message
- Each call generates a unique `id` via `crypto.randomUUID`

**Subscription (`src/hooks/useGameSubscription.test.ts`) — 4 new tests:**
- `chat` event handler is bound on the private-game channel
- Chat events are forwarded through `callbacksRef`
- Chat events use the updated ref when the callback changes
- `null` `pusherClient` and `null` `playerId` both skip subscription

## Summary by CodeRabbit

* **New Features**
  * Real-time in-game chat with message history and unread badges
  * Configurable chess clocks and time-control selection (Bullet/Blitz/Rapid)
  * Play vs. computer with difficulty selector, engine status indicator, and Stockfish-powered moves
  * Smart hint system with visual board highlights and hint UI (button/toast)
  * Chess clock display and improved game status panels; 2D/3D view toggling and new controls

* **Tests**
  * Extensive new unit and end-to-end test coverage for chat, clock, engine, hooks, and multiplayer flows

### CODE REVIEW

**Comment by ragaeeb** [comment]
# Chessie Code Review (gemini-3.1-pro-review)

Here is a comprehensive code review of the outgoing changes in the current branch against `main`, focusing on bugs, security, architecture, performance, developer experience, and testing.

## 🔴 CRITICAL / BUGS

### 1. `useViewPreference.ts` - Unsafe `localStorage` Access
**File:** `src/hooks/useViewPreference.ts`
**Issue:** While `typeof window !== 'undefined'` is checked, `localStorage.getItem()` and `localStorage.setItem()` are called without a `try-catch` block. In browsers with strict privacy settings (like Brave or Safari in certain modes), accessing `localStorage` throws a synchronous DOMException, which will crash the entire React application.
**Recommendation:** Wrap `localStorage` access in `try-catch` blocks, similar to what is done in `src/lib/playerIdentity.ts`.

### 2. Hardcoded Promotion in Computer Game
**File:** `src/app/game/computer/page.tsx`
**Issue:** `handleLocalMove` always promotes to a Queen (`promotion: 'q'`). If a user needs an underpromotion (e.g., to a Knight to deliver a smothered mate or avoid stalemate), they are unable to do so against the computer.
**Recommendation:** Implement a promotion prompt for the user, or provide a UI mechanism to choose the promotion piece.

### 3. Biome Linting Failure (Tailwind Directives)
**Files:** `src/app/globals.css`, `biome.json`
**Issue:** `bun run lint` fails because Biome cannot parse Tailwind-specific syntax like `@apply` or `@theme` in `src/app/globals.css` out of the box.
**Recommendation:** Configure Biome to support Tailwind CSS (e.g., adding cssModules configuration) or refactor the CSS file to standard utility classes in `src/app/layout.tsx`.

## 🟠 ARCHITECTURE & CODE SMELLS

### 1. The "God Component" Persists
**File:** `src/app/game/[gameId]/page.tsx`
**Issue:** Despite extracting some logic, the `GamePage` component is still massive (nearly 400 lines) and handles too many concerns: API calls, Pusher subscriptions, local rendering state, UI overlays, and game logic. 
**Recommendation:** Extract custom hooks like `useGameState`, `useGameNetwork`, and `useGameActions` as originally suggested in previous reviews to significantly improve maintainability.

### 2. Duplicated Computer Move Logic
**File:** `src/app/game/computer/page.tsx`
**Issue:** The simple AI logic (random move with capture priority) is baked directly into a `useEffect` inside the component. This is hard to test and tightly coupled to the UI.
**Recommendation:** Extract the computer AI logic into a separate module (e.g., `src/lib/bot.ts`) or hook (`useComputerOpponent`).

### 3. Mixed Concerns in `GameStatusPanel.tsx`
**File:** `src/components/GameStatusPanel.tsx`
**Issue:** The component now accepts `is2D` and `gameId` to handle view switching and routing. This tightly couples a presentational "status panel" to Next.js routing.
**Recommendation:** Pass an `onViewToggle` callback from the parent (`GamePage` and `ComputerGameContent`) instead of passing `gameId` and calling `router.replace` directly inside the panel.

## 🟡 PERFORMANCE & REACT BEST PRACTICES

### 1. Missing Memoization for Board Squares
**Files:** `src/hooks/useBoardElements.tsx` and `src/hooks/useBoardElements2D.tsx`
**Issue:** React is constantly re-rendering all 64 squares for any state change because the individual square components (`BoardSquare` and `BoardSquare3D`) are newly instantiated or mapped within `useMemo` but lack component-level memoization. 
**Recommendation:** Extract `BoardSquare` to a named component and wrap it comprehensively with `React.memo()`.

### 2. Extraneous Rerenders on `handleSquareClick`
**File:** `src/hooks/useBoardElements2D.tsx`
**Issue:** The `handleSquareClick` function depends on `gameStatus`, `playerColor`, etc., causing it to be re-created frequently and busting the `useMemo` cache for the board elements.
**Recommendation:** Use `useRef` for values like `gameStatus` or `playerColor` that are only read during the click event to avoid invalidating the callback and memoized elements on trivial state updates.

## 🟢 DEVELOPER EXPERIENCE (DX) & DOCUMENTATION

### 1. `biome-ignore` Leftovers
**File:** `src/app/game/computer/page.tsx`
**Issue:** There is a leftover `/* biome-ignore lint/a11y/useValidAriaRole: Custom component prop */` comment attached to `GameStatusPanel`. 
**Recommendation:** Review if `role="white" | "black"` clashes with actual HTML aria-role attributes, and rename to `playerRole` if so, removing the need for suppression.

### 2. Missing JSDoc Comments
**Files:** `src/shared/api.ts`, `netlify/functions/utils/rateLimit.ts`
**Issue:** Critical shared utility functions like `checkRateLimit`, `signPlayerId`, and the new API types lack JSDoc/inline comments explaining their purpose or the caching behavior of memory/Redis fallbacks.
**Recommendation:** Add detailed JSDoc comments to document these new security and rate-limiting utilities.

## 🔵 TESTING GAPS (Unit, Integration & E2E)

### 1. Unit Tests to Add
*   **`netlify/functions/utils/session.test.ts`**: Test `signPlayerId` and `verifyPlayerSignature` to ensure signatures match for valid inputs, maintain length, and fail for tampered/invalid signatures.
*   **`netlify/functions/utils/rateLimit.test.ts`**: Test the new rate limit functionality, specifically falling back to the memory store when Redis is unavailable, bucket expirations, and verifying limits enforce appropriately.
*   **`src/hooks/useViewPreference.test.ts`**: Test that the `is2D` preference correctly extracts from the URL query params, falls back to `localStorage`, defaults safely on SSR, and handles Storage quota exceptions properly.
*   **`src/lib/bot.test.ts`** (if extracted): Fast unit test to verify computer move generation correctly prioritizes immediate recaptures over random quiet moves.

### 2. Integration Tests
*   **`move.test.ts` & `pusherAuth.test.ts` Updates**: The current integration tests cover basic flows but need expanded testing for the new `playerSig` verification pipeline. Explicitly verify that payload spoofing with an incorrect or missing `playerSig` results in a secure `401 Unauthorized` block.
*   **Rate Limiting Integration**: Write tests for simulated API spam (triggering 31+ `/move` events or 61+ `/pusherAuth` events in a single mock interval) to verify a stable `429 Too Many Requests` envelope.

### 3. E2E Tests (Playwright/Cypress)
The project continues to lack robust E2E test coverage for multiplayer semantics.
*   **Multiplayer E2E Sync Flow test**:
    *   Initialize two browser instances (Context A, Context B).
    *   Context A creates a game and extracts the `gameId`.
    *   Context B navigates to `/game/{gameId}`.
    *   Context A executes a move (Pawn e2 -> e4); verify Context B's UI reflects the update in realtime via WebSockets.
    *   Test disconnection logic (kill Context B tab, verify Context A gets opponent left notice).
*   **View Navigation E2E**: Test switching between 2D and 3D views. Verify that the URL updates with `?view=2d` and `localStorage` retains the selection accurately across hard refreshes.
*   **Computer Match E2E**: Test a basic full game against the `/game/computer` endpoint, waiting for its delayed artificial response timing.

**Comment by ragaeeb** [comment]
# Claude Opus 4.6

1. **Re-enabled React StrictMode** (`next.config.ts`) — All three reviews flagged this. It was disabled for debugging and should remain on to catch effect cleanup issues.

2. **Fixed `useChessClock` rAF cleanup bug** — When `activeColor` changed between truthy values (e.g. `'white'` -> `'black'`), the old animation loop wasn't cancelled before starting a new one. Now any existing rAF is cancelled at the top of the effect, regardless of the new value.

3. **Shared `CHAT_MAX_LENGTH` constant** — The `MAX_LENGTH = 200` in `chat.ts` and `MAX_CHARS = 200` in `ChatPanel.tsx` were independent magic numbers. Extracted `CHAT_MAX_LENGTH` into `src/types/chat.ts` and both files now import from it.

4. **Hardened `getClientIp`** — Extracted into `netlify/functions/utils/http.ts` as a shared utility (was duplicated in `chat.ts` and `move.ts`). In production, only the trusted Netlify header `x-nf-client-connection-ip` is used. The spoofable `x-forwarded-for` fallback is restricted to local dev.

5. **Fixed inline type import** in `useGameNetwork.ts` — Replaced `import('@/types/clock').ClockState` with a proper module-level import.

6. **Deleted `chat_proposal.md`** — Implementation is complete; the design doc doesn't belong in the repo.

7. **Updated `AGENTS.md`** — Added documentation for the new chat architecture.

---

## Disagreed with (not fixed)

- **XSS sanitization for chat** (kimi #5, minimax #7) — React's `{msg.text}` renders as escaped text content. This is not an XSS vector. Both reviews acknowledge React handles this.

- **DebugPanel in production** (kimi #10, minimax #10) — Intentional user-facing diagnostic feature, hidden by default, ~2KB impact.

- **Logger module-level state** (kimi #11, minimax #17) — Standard singleton pattern. React context would be overkill for a cross-cutting utility.

- **Missing error boundaries** (kimi #9, minimax #6) — Valid suggestion but pre-existing architectural concern, not introduced by recent features.

- **`useCallback` for `tick` in `useEffect`** (minimax #16) — Defining the function inside `useEffect` is the correct React pattern for proper closure capture.

- **Ref mutation as "anti-pattern"** (minimax #14) — This is the standard pattern we deliberately adopted to prevent Pusher subscription storms.

- **Chat panel virtualization** (kimi #7, minimax #18, #21) — With a 200-message cap and simple text items, React handles this without performance issues.

- **JSDoc on everything** (all three) — Per project conventions, we avoid redundant comments. Type signatures are self-documenting.

- **Unused `Suspense` import** (minimax #12) — Factually incorrect. `Suspense` is used at line 393 of `computer/page.tsx`.

- **Test script complexity** (kimi #18, minimax #33) — Necessary workaround for Bun's `mock.module` leaking between test files in the same process.


1. **P0 — Session impersonation vulnerability** (`session.ts`) — `issuePlayerSession` now rejects (returns `null`) when a session already exists but the provided token doesn't match, instead of minting a new token that overwrites the legitimate player's session.

2. **P1 — Unvalidated `timeControl` input** (`move.ts`) — Added `isValidTimeControl()` that validates incoming time control values against the known `TIME_CONTROLS` presets. Arbitrary shapes/negative values/NaN are rejected with a `400`.

3. **P2 — False opponent disconnect for spectators** (`useGameSubscription.ts`) — The `member_removed` handler now checks `member.info.role !== 'spectator'` before firing `onOpponentDisconnected`. Spectator leaves no longer trigger false disconnect warnings.

4. **P2 — 3D board guard inconsistency** (`useBoardElements.tsx`) — Changed from `gameStatus === 'waiting-opponent'` to `gameStatus !== 'started'` to match the 2D board, blocking interactions in `not-started` state too.

5. **P2 — Hint quota not reset on New Game** (`useHint.ts`, `computer/page.tsx`) — Added `resetHints()` to `useHint` and call it in `handleNewGame`, so hint count resets per game, not per page session.

6. **`MoveBroadcastPayload` type drift** (`useGameSubscription.ts`) — Added optional `clock?: ClockState` to match what the server actually broadcasts.

7. **P3 — Test convention drift** (`netlifyRoutes.test.ts`) — Converted `test(...)` to `it('should ...')` per AGENTS.md convention.

8. **P3 — Leftover reasoning comments** (`move.test.ts`) — Removed ~30 lines of AI-generated exploratory commentary.

9. **Bonus — `pusherAuth.ts` IP hardening** — Replaced the inline `getClientIp` with the shared hardened utility from `utils/http.ts` (same fix we applied to `chat.ts` and `move.ts` earlier).

---

## Disagreed with (not fixed)

- **P1 — Timed games not server-authoritative** — Valid long-term concern, but this is a deliberate MVP design. Server-side timeout enforcement requires a background process or scheduler (Netlify Functions are request-driven and can't run a timer). Clock starting on first move (not game creation) is also standard for casual online chess. This needs a design discussion involving infrastructure changes (e.g. Pusher delayed events, external cron), not a code fix.

- **P1 — Unit test suite broken** — Already fixed in the previous conversation turn by extracting `ChatManager` to eliminate `mock.module('react')`.

- **P2 — Engine timeout/cancellation** — Valid concern but pre-existing architecture, not introduced by recent changes. Would need an `AbortController`-like pattern threaded through the Stockfish worker communication. Good follow-up issue.

- **P3 — Test scripts brittle** — The explicit file list is a necessary workaround for Bun's `mock.module` leaking between files in the same process. Already simplified by folding `chatManager.test.ts` into the main batch.


1. **Duplicate test scripts** (`package.json`) — `test` now delegates to `test:unit` via `bun run test:unit`.

2. **DebugPanel in production** (`layout.tsx`) — Now conditionally rendered only when `NODE_ENV !== 'production'`.

3. **Misleading clock test title** (`clock.test.ts`) — Renamed "should add increment to the moving player" to "should deduct elapsed time without adding increment" since the test uses a zero-increment time control.

4. **DebugPanel stale log count** (`DebugPanel.tsx`) — Added `setLogCount(logger.getLogCount())` before creating the interval so the count is accurate immediately on enable.

5. **Stale engineState in useStockfish** (`useStockfish.ts`) — When `difficulty` is cleared, now explicitly resets `engineState` to `'idle'` and clears refs.

6. **Brittle `toHaveBeenCalledTimes(2)`** (`chat.test.ts`) — Replaced with filtering for `SocketEvents.CHAT` calls specifically and asserting `toHaveLength(1)`.

7. **Engine getBestMove timeout** (`engine.ts`) — Added a 30s `setTimeout` that rejects the promise and transitions to `'error'` state if no `bestmove` arrives. The timeout is cleared on successful resolution.

8. **`getLogs()` returns mutable reference** (`logger.ts`) — Now returns `[...entries]` (defensive copy).

9. **Fragile `playerColor?.[0]` pattern** — Extracted `colorInitial()` and `isPlayerTurn()` helpers into `src/shared/player.ts`. Updated all 4 call sites (`page.tsx`, `GameStatusPanel.tsx`, `useBoardElements.tsx`, `useBoardElements2D.tsx`).

10. **Chat `parseBody` trusts JSON types** (`chat.ts`) — Added `isChatRequestBody` runtime type guard that validates all 5 fields are strings before casting. Prevents `text.trim()` throwing on non-string input.

11. **Unhandled Pusher trigger throw** (`chat.ts`) — Wrapped in try/catch, returns `502` with error message on failure.

12. **Chat test asserting 401 instead of 404** (`chat.test.ts`) — The test "should return 404 when gameId does not exist" was sending `playerSig: 'any'` which failed auth first. Now uses valid credentials so it actually reaches the game lookup path.

13. **3 chat tests asserting "Missing" message** (`chat.test.ts`) — Updated to `toBeDefined()` since the stricter type guard now rejects incomplete payloads earlier as "Invalid request body".

14. **`timeControl` validation** — Already implemented in the prior review round. Verified the `isValidTimeControl` function is present and working.

---

## Disagreed with (not fixed)

- **Server-side clock timeout on move** — Same position as before: Netlify Functions are request-driven with no background process capability. Enforcing timeout requires a scheduler or cron-like mechanism that this infrastructure doesn't support. Clock starting on first move is standard for casual online chess. Tracked as an architecture discussion, not a code fix.

- **GameRecord normalization for Redis migration** — The `timeControl` and `clock` fields were introduced alongside the game creation code that always populates them. There are no pre-existing Redis records without these fields — the feature launched as a unit. Adding a normalization layer adds complexity for a migration scenario that doesn't exist.

- **E2E test hardening** — The suggested changes (`page2.reload()`, `p2MoveCount` counter) add complexity to tests that are already mock-heavy. The E2E tests verify the UI integration with mocked network responses; the server-side validation logic is thoroughly covered by unit tests. The suggested changes also introduce timing sensitivities (`await expect.poll(...)`) that can cause flakiness.

- **`clock.ts` negative time logging** — The `Math.max(0, ...)` clamping is intentional and correct. Adding logging for "clock drift" or "flagged earlier" is a monitoring concern that belongs in application-level observability, not in a pure utility function.


| # | File | Finding | Verdict |
|---|------|---------|---------|
| 1 | `netlify/functions/utils/session.ts` | `issuePlayerSession` used plain `===` comparison instead of `timingSafeEqual` | **Fixed** — now uses `timingSafeEqual` via `Buffer`, consistent with `verifyPlayerSignature`. Also handles `!providedToken` separately before buffer comparison. |
| 2 | `src/components/HintToast.tsx` | Dismiss button missing accessible name | **Fixed** — added `aria-label="Dismiss hint"` to the icon-only button. |
| 3 | `src/hooks/useChessClock.ts` | `lastTickRef` set from `serverState.lastTickAt` could cause negative elapsed due to clock skew | **Fixed** — `lastTickRef.current` now always set to `Date.now()` (client time) on sync, and elapsed is clamped with `Math.max(0, ...)`. |
| 4 | `src/hooks/useHint.ts` | `requestHint` could be entered concurrently before React re-renders the `loading` state | **Fixed** — added synchronous `inFlightRef` guard, set immediately before async work and cleared in `finally`. Removed `loading` from the dependency array since the ref replaces it as the guard. |
| 5 | `src/hooks/useStockfish.ts` | `getBestMove` re-read refs after `await`, risking stale engine access | **Fixed** — pinned `engineRef.current` and `initPromiseRef.current` to local consts before awaiting; added a staleness check after `await initPromise`. |
| 6 | `src/lib/chatApi.test.ts` | `globalThis.crypto` overridden in `beforeEach` but never restored | **Fixed** — saved `originalCrypto` and restored it in `afterEach`, along with `uuidMock.mockReset()` + re-implementation of default. |
| 7 | `src/lib/stockfish.ts` | Concurrent `getBestMove` calls overwrite the shared `onmessage` handler | **Fixed** — added module-level `requestInFlight` guard that rejects if a request is already pending. All exit paths (success, timeout, no-move error) go through a `cleanup()` function. `terminateStockfish` also resets the flag. |
| 8 | `src/lib/stockfish.ts` | `typeof e === 'string'` is wrong — `e` is always a `MessageEvent` | **Fixed** — changed to `typeof e.data === 'string' ? e.data : ''` with an early return for empty/non-string data. |

All 163 tests pass, linting is clean (pre-existing warnings only), and the build succeeds.


**Major fixes:**

| Issue | Fix |
|---|---|
| **Clock authority — unlimited first-move time** | Clock now activates for white (`activeColor: 'white'`, `lastTickAt: now`) when the second player joins, in `respondForJoinDecision`. White's clock ticks from game start, not from first move. |
| **Clock authority — no pre-move timeout** | New `checkClockExpiry` helper in `move.ts` computes server-side elapsed time and rejects moves with `400 Time expired` + broadcasts `GAME_OVER` with `reason: 'timeout'` when a player's clock has run out. |
| **Clock shape drift in Redis** | Added `normalizeGameRecord` in `gameStore.ts` that restores `clock.activeColor` and `clock.lastTickAt` to `null` when Lua's `cjson.encode` strips them. Applied to both `getGameRecord` and `ensureGameRecord`. |
| **CI test stability** | Added `test:ci` script that mirrors `test:unit` but includes `--coverage --coverage-reporter=lcov` flags. Updated both CI jobs in `build.yml` to use `bun run test:ci` instead of bare `bun test --coverage`. |

**Architecture / DX / Testing fixes:**

| # | Issue | Fix |
|---|---|---|
| 1 | **GameStatusPanel coupled to routing** | Replaced `gameId` prop with `onViewToggle: () => void` callback. `PanelControls` no longer imports `useRouter`. The parent page (`[gameId]/page.tsx` and `computer/page.tsx`) owns the route change. |
| 2 | **Missing JSDoc** | Added JSDoc to `checkRateLimit` (documents Redis/memory fallback, parameters) and `JoinGameResponse` / `MoveAction` in `shared/api.ts` (documents session token schema, clock semantics). |
| 3 | **Missing unit tests** | Created `session.test.ts` (11 tests: token minting, idempotent re-auth, impersonation rejection, signature verification, length-mismatch rejection) and `rateLimit.test.ts` (7 tests: first request allowed, decrement tracking, limit exceeded, undefined IP normalization, CSV IP normalization, action isolation, IP isolation). |

**Items from the attached reviews (minimax, kimi) — not fixed, with rationale:**

- **Stockfish worker memory leak on navigation** (minimax #1): The singleton worker is intentionally module-scoped. In Next.js client-side navigation the module stays loaded; full page navigations already GC the worker. Adding cleanup effects risks premature termination mid-computation.
- **Error boundaries** (minimax #2, kimi #4): Pre-existing architectural concern, not introduced by recent changes. Agreed it's useful but it's a separate feature.
- **Missing hook-level tests** (minimax #3 — `useChessClock.test.ts`, `useHint.test.ts`): The core logic of these hooks is simple state wiring. The clock math is tested in `clock.test.ts`, the hint system via `stockfish.test.ts`. Adding React-rendering tests adds complexity with minimal coverage gain.
- **Duplicate game state** (minimax #4): Acknowledged design concern; the page is a composition root, and the state pieces have different lifecycles.
- **Reconnection UI** (minimax #5): A separate feature, not a bug introduced by recent changes.
- **ChatManager mutation pattern** (kimi #1): Low risk given the `bump()` re-render trigger. The pattern is intentional for avoiding unnecessary object allocation.
- **stockfish.ts module-level state** (kimi #2): Standard singleton pattern; the `requestInFlight` guard we just added actually makes it safer.
- **XSS sanitization server-side** (kimi #3): React auto-escapes JSX text. Server-side sanitization for defense-in-depth is reasonable but low priority since the only consumer is React.

File: src/lib/stockfish.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+export function getBestMove(fen: string, depth = 10): Promise<BestMove> {
+    return new Promise((resolve, reject) => {
+        const sf = getWorker();
+        const timeout = setTimeout(() => reject(new Error('Stockfish timed out')), 5000);
+
+        sf.onmessage = (e: MessageEvent) => {
+            const line: string = typeof e === 'string' ? e : e.data;
+
+            if (line.startsWith('bestmove')) {
+                clearTimeout(timeout);
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/lib/stockfish.ts]: https://github.com/ragaeeb/chessie/pull/21/commits/5fe64384127bd36a809f42e3381f6847ac449377

File: netlify/functions/chat.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        return respond(400, { error: 'Missing required fields' });
+    }
+
+    if (!(await verifyPlayerSignature(playerId, playerSig))) {
+        return respond(401, { error: 'Invalid player session' });
+    }
+
+    const trimmed = text.trim();
+    if (!trimmed || trimmed.length > MAX_LENGTH) {
+        return respond(400, { error: `Message must be 1–${MAX_LENGTH} characters` });
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply netlify/functions/chat.ts]: https://github.com/ragaeeb/chessie/pull/21/commits/5fe64384127bd36a809f42e3381f6847ac449377

File: e2e/multiplayer.e2e.ts [hunk 3]
```diff
@@ -274,6 +270,140 @@ test.describe('/game/[gameId] - Multiplayer Game', () => {
...
+                            role: 'black',
+                            color: 'black',
+                            status: 'active',
+                            fen: currentFen,
+                            playerSig: BLACK_SIG,
+                        }),
+                    });
+                } else if (body.action === 'move') {
+                    currentFen = AFTER_E4_E5_FEN;
+                    await route.fulfill({
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply e2e/multiplayer.e2e.ts]: https://github.com/ragaeeb/chessie/pull/21/commits/5fe64384127bd36a809f42e3381f6847ac449377

File: src/app/game/computer/page.tsx [hunk 4]
```diff
@@ -2,28 +2,192 @@
...
+        </motion.div>
+    );
+}
+
+function parseUCIMove(uci: string): { from: string; to: string; promotion?: string } {
+    const from = uci.slice(0, 2);
+    const to = uci.slice(2, 4);
+    const promotion = uci.length > 4 ? uci[4] : undefined;
+    return { from, to, promotion };
+}
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/app/game/computer/page.tsx]: 3d55fe607ce10311b8aa0445f77d359ab73f62d2

File: src/app/game/[gameId]/page.tsx [hunk 5]
```diff
@@ -56,142 +70,156 @@ const GamePage: React.FC = () => {
...
+    const { white: clockWhite, black: clockBlack } = useChessClock(serverClock);
+    const getFen = () => game.fen();
+    const {
+        hint,
+        hintSquares,
+        hintsRemaining,
+        loading: hintLoading,
+        requestHint,
+        clearHint,
+    } = useHint(getFen, playerColor);
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/app/game/[gameId]/page.tsx]: 3d55fe607ce10311b8aa0445f77d359ab73f62d2
